arangod/Pregel/Graph.h

```diff
@@ -68,54 +67,60 @@ class Edge {
   template <typename V, typename E2>
   friend class GraphStore;
 
+  static_assert(sizeof(std::string) > 2, "");
+  char* _toKey;             // uint64_t
+  size_t _toKeyLength;      // uint64_t
   // PregelShard _sourceShard;
-  PregelShard _targetShard;
-  PregelKey _toKey;
+  PregelShard _targetShard; // uint16_t
+
   E _data;
 
  public:
   // EdgeEntry() : _nextEntryOffset(0), _dataSize(0), _vertexIDSize(0) {}
-  Edge() : _targetShard(InvalidPregelShard), _data(0) {}
-  Edge(PregelShard target, PregelKey const& key)
-      : _targetShard(target), _toKey(key), _data(0) {}
+//  Edge() : _targetShard(InvalidPregelShard), _data(0) {}
+//  Edge(PregelShard target, PregelKey const& key)
+//      : _targetShard(target), _toKey(key), _data(0) {}
 
   // size_t getSize() { return sizeof(EdgeEntry) + _vertexIDSize + _dataSize; }
-  PregelKey const& toKey() const { return _toKey; }
+  StringRef toKey() const { return StringRef(_toKey, _toKeyLength); }
   // size_t getDataSize() { return _dataSize; }
-  inline E* data() {
-    return &_data;  // static_cast<E>(this + sizeof(EdgeEntry) + _vertexIDSize);
+  E& data() {
+    return _data;  // static_cast<E>(this + sizeof(EdgeEntry) + _vertexIDSize);
   }
-  // inline PregelShard sourceShard() const { return _sourceShard; }
-  inline PregelShard targetShard() const { return _targetShard; }
+  // PregelShard sourceShard() const { return _sourceShard; }
+  PregelShard targetShard() const { return _targetShard; }
 };
 
-class VertexEntry {
-  template <typename V, typename E>
-  friend class GraphStore;
-
-  PregelShard _shard;
-  PregelKey _key;
-  size_t _vertexDataOffset = 0;
-  size_t _edgeDataOffset = 0;
-  size_t _edgeCount = 0;
-  bool _active = true;
+template <typename V, typename E>
+class Vertex {
+  friend class GraphStore<V,E>;
+  
+  const char* _key; // uint64_t
+  size_t _keyLength; // uint64_t
```

Comment:
Can we make `_keyLength` a uint16_t? I can't think of any keys longer than 255 chars anyway, and by shrinking the type for _keyLength, we can move it further down and reduce the size of the struct after padding.
We can then make all accesses to the `_keyLength` attribute go through an inline function that returns the size as a `size_t`.